[checker] Refactor by sharing code with the kernel

maximedenes opened the pull request:
For historical reasons, the checker was duplicating a lot of code of the
kernel. The main differences I found were bug fixes that had not been
backported.

With this patch, the checker uses the kernel as a library to serve the
same purpose as before: validation of a `.vo` file, re-typechecking all
definitions a posteriori.

We also rename some files from the checker so that they don't clash with
kernel files.

Based on #8684 [merged].

EDIT: now depends on #8842 [merged] and #8870 [merged].

Still TODO:
- [x] Fix the Dune build
- [x] Make sure we check the same things as before on modules (@ppedrot can you have a look?)
- [X] Disable VM (and native) for conversions triggered by the checker
- [X] Make sure the checker can't load plugins, as claimed in the manual

ejgallego commented:
Note @maximedenes the last commit. IMHO worth it, bringing in half of `library` for a meager 20 lines seems like serious stuff, but review and do as you prefer.

SkySkimmer commented:
This depends on #8684 IIUC.

SkySkimmer commented:
@ejgallego did you forget to remove the -I library from makefile in your commit or is it still somehow needed?

SkySkimmer commented on checker/checkTypes.ml:
```diff
@@ -0,0 +1,36 @@
+      | None::pl,d::params -> check_p (push_rel d env) pl params
+      | [], _ -> ()
+      | _ -> failwith "check_poly: not the right number of params" in
+  check_p env pl (List.rev params)
```
Does this really deserve a file? It could live in kernel/typeops I think.

SkySkimmer commented on checker/mod_checking.ml:
```diff
@@ -89,40 +102,41 @@ and check_structure_field env mp lab res = function
-  | MEwith _ -> error_with_module ()
+      let farg_id, farg_b, fbody_b = Modops.destr_functor sign in
+      let mtb = Modops.module_type_of_module (lookup_module mp env) in
+      let _ = check_subtypes env mtb farg_b in (* Should we stop inferring constraints? *)
```
Shouldn't we check that the constraints are satisfied instead of dropping them?

maximedenes replied:
Right, what function should I call?

maximedenes replied:
Although, isn't it already checked by the kernel at subtyping time? If not, when does the kernel check it?

SkySkimmer replied:
The kernel adds the constraints to the environment. This part corresponds to some subcall of translate_module so the constraints are first added to the module_body that function produces, then when the module is added to the safe env the constraints are added.
When checking use Environ.check_constraints (returns bool).

ppedrot commented:
Without presuming of the tradeoffs of this PR, I am a bit saddened by the disappearance of the `cic.mli` file that was a pretty good summary of the structures involved in the definition of the Coq kernel.

SkySkimmer commented on checker/safe_checking.ml:
```diff
@@ -25,6 +25,18 @@ let pr_dirpath dp = str (DirPath.to_string dp)
 let genv = ref empty_env
 let get_env () = !genv
```
We should carry this global ref functionally at some point.

maximedenes replied:
I'm taking care of this one, in fact almost the whole file can be refactored as well.

ejgallego commented:
> @ejgallego did you forget to remove the -I library from makefile in your commit or is it still somehow needed?

I just don't use use `Makefile`. The current make system is anyways not well suited for the purpose of a containerized build, unless you specify all the rules by hand, that is.

ejgallego commented:
> I'm refactoring more with library, but we can evaluate again the tradeoff for the final patch.

Certainly I am convinced that bringing in `global/summary/nametab` etc... must be avoided.

maximedenes commented:
> Certainly I am convinced that bringing in `global/summary/nametab` etc... must be avoided.

Oh, not at all! In fact, it is a bug of the current patch that we are not using the global environment and instead have a local one that is badly kept in sync. Let me push the fixed patch so that you see.

ejgallego commented:
> Oh, not at all! In fact, it is a bug of the current patch that we are not using the global environment and instead have a local one that is badly kept in sync. Let me push the fixed patch so that you see.

well let me remove `global` from that list, but certainly `nametab` and `summary` must not be used.

maximedenes commented:
> but certainly `nametab` and `summary` must not be used.

Indeed, I don't think they are used. In fact, `coqchk` is like `coqc`, it is a toplevel that doesn't need interactive features, and what is special is that it takes as input a `.vo` which it validates.

maximedenes commented:
@ejgallego do you have a backup of your last commit? I'm going to have to force push the larger refactoring which removes it (but keeps the one for the Dune build, which may have to be adapted).

ejgallego commented:
> @ejgallego do you have a backup of your last commit? I'm going to have to force push the larger refactoring which removes it (but keeps the one for the Dune build, which may have to be adapted).

Umm why do I need a backup of it then?

IMHO you should not push until you make Dune satisfied, not especially with the current build rules you are using. Otherwise, the PR is useless as we cannot see what the new checker is using.

Zimmi48 commented:
> Oh yes, I need to add code to ensure that, I forgot about it, thanks!

Wouldn't it be safer to do as Emilio suggested: do a kernel_lib from what is common to the checker and the kernel and use only this? This would avoid having to "disable" the VM and native compute, the module and section management, the plugins, etc. This would also avoid accidental violation of this invariant in future updates.

maximedenes commented:
> Wouldn't it be safer to do as Emilio suggested: do a kernel_lib from what is common to the checker and the kernel and use only this? This would avoid having to "disable" the VM and native compute, the module and section management, the plugins, the more efficient universe graph algorithm, etc. This would also avoid accidental violation of this invariant in future updates.

And how do I avoid the duplication of all the code from `library`, then?

SkySkimmer commented:
The checker should handle the environment functionally, not with library-style global refs.

maximedenes commented:
> The checker should handle the environment functionally, not with library-style global refs.

Sorry, but what is different between coqtop and coqchk, for this matter?

SkySkimmer commented:
coqtop has a ton of stuff on top of the kernel
coqchk doesn't have anything on top of the kernel

Gaëtan Gilbert

On 19/10/2018 17:01, Maxime Dénès wrote:
>     The checker should handle the environment functionally, not with
>     library-style global refs.
> 
> Sorry, but what is different between coqtop and coqchk, for this matter?
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub 
> <https://github.com/coq/coq/pull/8773#issuecomment-431392617>, or mute 
> the thread 
> <https://github.com/notifications/unsubscribe-auth/ACWQ7N3WYEVkD2DbqqT5NDJg11VmT5e6ks5umelAgaJpZM4Xvdp4>.
>

maximedenes commented:
> coqchk doesn't have anything on top of the kernel

that's not true, look at its library loading code...

SkySkimmer commented:
I may have exaggerated some but is it really that bad? If so I'm sorry 
for assuming but also a bit scared at the checker complexity that I 
wasn't aware of.

Gaëtan Gilbert

On 19/10/2018 17:05, Maxime Dénès wrote:
>     coqchk doesn't have anything on top of the kernel
> 
> that's not true, look at its library loading code...
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub 
> <https://github.com/coq/coq/pull/8773#issuecomment-431394009>, or mute 
> the thread 
> <https://github.com/notifications/unsubscribe-auth/ACWQ7Kw_zv4nXVmGwdl6ZJUF2kzIB0Lvks5umepAgaJpZM4Xvdp4>.
>

maximedenes commented:
> I may have exaggerated some but is it really that bad? If so I'm sorry for assuming but also a bit scared at the checker complexity that I wasn't aware of.

Well, it is not horrible, but it has dependency checking, checksum validation, logical paths mapping (I haven't refactored that part yet), etc.

Also, a lot of the added value code of the checker lives above the kernel, and validates the shape of `.vo` file. If you read the description of the checker in the manual, you'll realize it can be seen as an alternative toplevel which takes `vo` files as input and validates them.

I strongly believe that code duplication should be removed as much as possible. But I understand it would be better to not bring too much of the codebase in the checker at the same time. I think @ejgallego 's remark go in the direction of doing these code imports in a fine-grained way, so that we get the best of both worlds, maybe at the cost of reorganizing a bit the codebase.

The current patch is suboptimal in that respect, and I'll try to improve it once I fix the last remaining (universe-related) bugs.

maximedenes commented:
> maybe it is worth renaming the checker so that it is clear that its role is not the one it used to be (whatever that was really doing or perceived to be doing) but rather to perform a "vo roundtrip validation".

I'm not sure I understand this comment. As @Zimmi48 mentioned, the role of the checker was rather precisely described in the manual and is still the same as before. I don't really know what else was perceived or where it was coming from, in particular I doubt the name played a big role there.

maximedenes commented:
> I'm not sure I understand this comment. As @Zimmi48 mentioned, the role of the checker was rather precisely described in the manual and is still the same as before. I don't really know what else was perceived or where it was coming from, in particular I doubt the name played a big role there.

But of course, I'm not opposed to a change of name for the checker, but it is fully orthogonal to this PR which does not change its role at all.

gares commented:
Well, the other day you told me that some people were thinking it was an independent checker, as in de bruijn criterium. If you think there is no confusion after all, then I'm fine with the name.

maximedenes commented:
> Well, the other day you told me that some people were thinking it was an independent checker, as in de bruijn criterium. If you think there is no confusion after all, then I'm fine with the name.

Yeah, I've heard that before, but I'm not sure it comes from the name. Honestly, I don't really know where the confusion comes from. Maybe simply the fact that Bruno was involved in both Coq-in-Coq and the checker.

ejgallego commented:
After another look it should be possible to have a checker with minimal duplication.

Proper organization could require virtual libraries (thus not possible today) or defining new interfaces, but on the other the amount of API we are talking about is not a lot.

ejgallego commented on dev/ci/docker/bionic_coq/Dockerfile:
```diff
@@ -37,7 +37,7 @@ ENV COMPILER="4.05.0"
 # `num` does not have a version number as the right version to install varies
 # with the compiler version.
-ENV BASE_OPAM="num ocamlfind.1.8.0 dune.1.2.1 ounit.2.0.8 odoc.1.3.0" \
+ENV BASE_OPAM="num ocamlfind.1.8.0 dune.1.4.0 ounit.2.0.8 odoc.1.3.0" \
```
You should not need to bump the docker image anymore, 1.4.0 is in `master` now.

maximedenes commented:
This is ready for last reviews / merge. @ppedrot can you be the assignee? Thanks!

maximedenes commented:
Oh no, there is still two  FIXMEs, will address them now, sorry.

ejgallego commented on checker/dune:
```diff
@@ -1,7 +1,31 @@
+  %{project_root}/kernel/{subtyping,term_typing,safe_typing,entries,cooking}.ml{,i})
+
+(copy_files
+  %{project_root}/library/{decl_kinds,declaremods,summary,libnames,nametab,loadpath,globnames,libobject,lib,library}.ml{,i})
```
As discussed in the comment and in the branch, I think this should be considerably thinned.

maximedenes commented:
@ejgallego I addressed your comments, the checker now depends only on kernel files. We may want to refactor the library code similarly, but I'll leave that for another PR.

ejgallego commented on checker/dune:
```diff
@@ -1,7 +1,25 @@
+(copy_files
+  %{project_root}/kernel/{subtyping,term_typing,safe_typing,entries,cooking}.ml{,i})
+
+; VM stuff
```
If you are disabling vm compute do you still to link this?

maximedenes replied:
Good question, I thought we could provide a flag that enables VM / native for the checker, for cases when checking takes too much time otherwise. But I didn't expose such a flag yet (the machinery is there, though).

Do you think I should remove the runtime?

ejgallego replied:
I guess it is fine to leave it as is for now. Would you refine the VM/Native hooks indeed you could remove from the build a lot of files, which could be important for certification/review purposes, but that can happen later I guess.

ppedrot commented:
Last round of complains / comments.

SkySkimmer commented on Makefile.checker:
```diff
@@ -34,22 +34,8 @@ CHECKMLLIBFILE := checker/.mllibfiles
-
 ifeq ($(BEST),opt)
-$(CHICKEN): config/config.cmxa checker/check.cmxa checker/coqchk.mli checker/coqchk.ml
+$(CHICKEN): $(CORECMA:.cma=.cmxa) checker/check.cmxa checker/coqchk.mli checker/coqchk.ml
```
Why are we linking all CORECMA here?

ejgallego replied:
Indeed, this is a leftover of the original PR that used many files around the codebase. This should be able to be changed to `kernel.cma` now. @ppedrot wait until this is fixed please.

maximedenes replied:
Fixed.

SkySkimmer commented:
It's somehow strange to me that in make we link kernel.cma but in dune we copy source files around. Why's it done this way?

maximedenes commented:
@ejgallego wanted to do it that way in Dune. The Makefile stuff can be seen as legacy anyway, although it also does not seem bad to me to link with libraries as opposed to individual files.

SkySkimmer commented:
I guess it needs `$(LIBCOQRUN)`

maximedenes commented:
> Dependencies in make are wrong, if I `make bin/coqchk` from clean state I get

Fixed, thanks for noticing!

SkySkimmer commented on Makefile.checker:
```diff
@@ -22,7 +22,7 @@ CHICKEN:=bin/coqchk$(EXE)
 # The sources
 
-CHKLIBS:= -I config -I clib -I lib -I checker
+CHKLIBS:= -I config -I clib -I lib -I checker -I kernel -I kernel/byterun -I library -I engine
```
There should be no need to `-I library -I engine` now.

maximedenes replied:
Right, it was a leftover, now removed.

Zimmi48 commented on checker/dune:
```diff
@@ -1,7 +1,25 @@
-(rule (copy %{project_root}/kernel/names.mli names.mli))
-(rule (copy %{project_root}/kernel/esubst.ml esubst.ml))
-(rule (copy %{project_root}/kernel/esubst.mli esubst.mli))
+(copy_files
```
Isn't `copy_files#` usually preferable? Cf. https://dune.readthedocs.io/en/latest/dune-files.html#copy-files

ejgallego replied:
Good question, this is a bit of a special case but indeed `copy_files#` seems to be better.

maximedenes replied:
I changed to `copy_files#`, thanks.

ppedrot commented:
*En avant Guingamp.*
